design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,433 @@
+`GET <conjur-ui-url>/ui/callback?sut=<sut>&code_verifier=<code_verifier>`.
+The Conjur UI then calls the `POST /authn-sut/<account>/<login>/authenticate`
+endpoint with the SUT and code verifier to authenticate the user to Conjur and
+retrieve a short-lived auth token. The Conjur UI then starts a session for the
+user.
+
+![PSM Integration Workflow](./psm.png)
+
+##### Log Messages
+
+##### Error Log level
+
+| | Scenario | Log message |
+|--- |-----------------------------------------------------------------------  |----------------------------------------------------------------------------------------|
+| 1  | Authenticator is not enabled (in DB/ENV)                                | Authenticator '{0-authenticator-name}' is not enabled                                                 |
+| 2  | Webservice is not defined in a Conjur policy                            | Webservice '{0-webservice-name}' not found                                          |
+| 3  | User or host is not permitted to authenticate with the webservice               | '{0-role-name}' does not have 'authenticate' privilege on {1-service-name}             |
+| 4  | User or host is not defined in Conjur                                           | '{0-role-name}' wasn't found                                                           |
+| 5  | Login error                               | Login error: ...                                     |
+| 6  | Missing SUT in /authenticate request body | Single use token not provided    |
+| 7  | SUT not found for user, or expired | Single use token is invalid    |
+| 8  | Missing code challenge in /login request body | Code challenge not provided    |
+| 9  | Invalid code challenge in /login request body | Code challenge not valid    |
+| 10 | Invalid code challenge algorithm in /login request headers | Code challenge algorithm not supported    |
+| 11 | Missing code challenge algorithm in /login request headers | Code challenge algorithm not provided    |
+| 12 | Missing code verifier in /authenticate request body | Code verifier not provided    |
+| 13 | Code verifier does not match code challenge | Code verifier does not match code challenge    |
+
+###### Debug Log level
+
+|    | Scenario                                              | Log message                                                                            |
+|--- |-------------------------------------------------------|----------------------------------------------------------------------------------------|
+| 1  | After validating SUT | Single use token for {0-role-name} has been validated successfully                |
+| 2  | Attempt to use expired SUT | Single use token for {0-role-name} has expired    |
+| 3  | Attempt to use used SUT | Single use token for {0-role-name} has already been used    |
+| 4  | Attempt to use SUT for different user | Single use token for {0-role-name} is not valid for {1-role-name}    |
```

Comment:
I wonder if this should be only a debug message. This might be an indicator of an active attack and thus something we should be surfacing to the audit DB and maybe even at a Warning log level.